runtime/gc_base/ScavengerForwardedHeader.cpp

```diff
@@ -48,33 +40,23 @@ MM_ScavengerForwardedHeader::validateAssumptions()
 omrobjectptr_t
 MM_ScavengerForwardedHeader::setForwardedObject(omrobjectptr_t destinationObjectPtr)
 {
-	/* class slot must be aligned to UDATA */
-	Assert_MM_true(0 == (offsetof(J9Object, clazz) % sizeof(UDATA)));
-
 	Assert_MM_false(isForwardedPointer());
 
-	volatile MutableHeaderFields* objectHeader = (volatile MutableHeaderFields *)_objectPtr;
-	UDATA oldValue = *(UDATA *)&_preserved.clazz;
-
-#if defined (OMR_GC_COMPRESSED_POINTERS) && !defined(J9VM_ENV_LITTLE_ENDIAN)
-	/*
-	 *  Forwarded tag should be in low bits of the pointer and at the same time be in class slot
-	 * To get it for compressed big endian just swap halves of pointer
-	 */
-	/* low half */
-	UDATA newValue = ((UDATA)destinationObjectPtr >> 32) & 0xffffffff;
-	/* add a high half */
-	newValue |= ((UDATA)destinationObjectPtr | FORWARDED_TAG) << 32;
-
-#else /* defined (OMR_GC_COMPRESSED_POINTERS) && !defined(J9VM_ENV_LITTLE_ENDIAN) */
-
-	/* little endian or not compressed - write UDATA bytes straight */
+	UDATA oldValue = _preserved;
 	UDATA newValue = (UDATA)destinationObjectPtr | FORWARDED_TAG;
 
-#endif /* defined (OMR_GC_COMPRESSED_POINTERS) && !defined(J9VM_ENV_LITTLE_ENDIAN) */
+#if defined(OMR_GC_COMPRESSED_POINTERS) && !defined(J9VM_ENV_LITTLE_ENDIAN)
+	if (compressObjectReferences()) {
+		/*
+		 * Forwarded tag should is in the low bits of the pointer, but must be in the class pointer
+		 * for purposes of testing. Endian swap the pointer to achieve this.
+		 */
+		newValue = (newValue >> 32) | (newValue << 32);
```

Comment:
There is an assumption that shift right operation is logical, not arithmetic. It is how it should be for unsigned, however we had problems with compilers in the past. I hope it should work properly for all compilers we are using now